File: 2020/d15/d15.rb
#!/usr/bin/env ruby

SEED = File.read(ARGV[0]).split(",").map(&method(:Integer))

class Game
  attr_reader :ages, :last_turn, :last_said, :say_next

  def initialize(seed)
    # number spoken => last turn it was spoken
    @ages = Hash[
      seed.each_with_index.map { |n, idx|
        [n, idx + 1]
      }
    ]
    @last_said = seed[-1]
    @say_next = 0 # no repeats in seed
    @last_turn = seed.count
  end

  def play_turn!
    # puts "DEBUG current_turn=#{last_turn + 1}, last_said=#{last_said} say_this_turn=#{say_next} ages=#{ages}"
    @last_said = say_next
    @last_turn += 1
    @say_next = last_turn - ages.fetch(say_next, last_turn)
    ages[last_said] = last_turn
  end

  def play_turns!(turn_count)
    until last_turn == turn_count
      play_turn!
    end
  end
end

# p1
g = Game.new(SEED)
g.play_turns!(2020)
puts "p1: after 2020 turns, last number spoken was #{g.last_said}"

# p2
g = Game.new(SEED)
g.play_turns!(30_000_000)
puts "p2: after 30M turns, last number spoken was #{g.last_said}"

File: 2020/d15/d15.jl
#!/usr/bin/env julia

mutable struct GameState
  ages::Dict{Int, Int}
  last_turn::Int
  last_said::Int
  say_next::Int
end

function load_seed()
  read(open(ARGS[1]), String) |>
    Base.Fix2(split, r","m) |>
    Base.Fix1(map, strip) |>
    Base.Fix1(map, Base.Fix1(parse, Int))
end

function seed_game(seed):: GameState
  ages = Dict()
  for i in eachindex(seed)
    ages[seed[i]] = i
  end

  GameState(ages, length(seed), last(seed), 0)
end

function take_turn(game::GameState)::GameState
  # println("DEBUG: last_turn=$(game.last_turn) last_said=$(game.last_said) say_now=$(game.say_next) ages=$(game.ages)")
  game.last_said = game.say_next
  game.last_turn += 1
  game.say_next = game.last_turn - get(game.ages, game.say_next, game.last_turn)
  game.ages[game.last_said] = game.last_turn

  game
end

function go_to_turn(game::GameState, turn_count::Int)::GameState
  while game.last_turn < turn_count
    take_turn(game)
  end

  game
end

# p1
g = seed_game(load_seed())
g = go_to_turn(g, 2020)
println("p1: after $(g.last_turn) turns, last said was $(g.last_said)")

# p2
g = seed_game(load_seed())
g = go_to_turn(g, 30_000_000)
println("p1: after $(g.last_turn) turns, last said was $(g.last_said)")
